### PR TITLE
Store logical replication origin in KV storage

### PR DESCRIPTION
**Pull request by knizhnik**
Store logical replication origin in KV storage

## Problem

See  #6977

## Summary of changes

* Extract origin_lsn from commit WAl record
* Add ReplOrigin key to KV storage and store origin_lsn
* In basebackup replace snapshot origin_lsn with last committed origin_lsn at basebackup LSN

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by jcsp** [comment]
This is for the "outbound logical replication" feature, right?

The high level decision on this is going to be whether we ship this feature before or after writing some more scalable KV store -- if we build that, then we'll want to store this state in there, to avoid the overhead of using a whole page for each value.

**Comment by knizhnik** [comment]
> This is for the "outbound logical replication" feature, right?
> 
> The high level decision on this is going to be whether we ship this feature before or after writing some more scalable KV store -- if we build that, then we'll want to store this state in there, to avoid the overhead of using a separate page for each value.

No,  it is for inbound logical replication (Neon node is subscriber).
The problem is that now restarted Neon instance can send from restart LSN and receive WAL records which are already applied, causing unique violation errors.

**Comment by jcsp** [comment]
@kelvich can you put this into context within the overall LR plan?  I would rather avoid merging a change like this if we're going to have to later write more code to migrate this content into a newer storage backend.

If we do want to merge a change like this, then obviously it's going to need tests, and clear limits on how many of these keys are going to get created (I do not want to merge a feature that creates an unbounded number of tiny pages)

**Comment by knizhnik** [comment]
> @kelvich can you put this into context within the overall LR plan? I would rather avoid merging a change like this if we're going to have to later write more code to migrate this content into a newer storage backend.
> 
> If we do want to merge a change like this, then obviously it's going to need tests, and clear limits on how many of these keys are going to get created (I do not want to merge a feature that creates an unbounded number of tiny pages)

Number of keys is limited by number of origins, and it is limited by maximal number of slots (10 by default).

File: libs/pageserver_api/src/key.rs
Comment (hlinnaka) [inline libs/pageserver_api/src/key.rs:232]: Let's not take this AuxFiles approach with a HashMap again. For all the same write amplification issues we had with auxfiles. Let's treat each replication origin as a separate kv-pair, and have one directory entry that lists them.

File: pageserver/src/basebackup.rs
Comment (hlinnaka) [inline pageserver/src/basebackup.rs:324]: This is just for backwards-compatibility with tenants that already started using logical replication, right? Let's mention it in the comment.

File: libs/pageserver_api/src/key.rs
Comment (knizhnik) [reply libs/pageserver_api/src/key.rs:232]: I have thought a lot about it, but I think that in this case this approach is preferable.

We in any case have to store directory for replorigins, right?
In case of replication origins content (origin_id+origin_lsn) have almost the same size as just reference (origin_id).
So I do not see much sense in storing origin_lsn separately. It will not reduce speed but significantly increase time (because we have to fetch all elements of directory). 

Also number of replorigins is limited by `max_slot_number`  - it is 120 by default and in most case there will be no more than just one slot and replorigin.

File: pageserver/src/basebackup.rs
Comment (knizhnik) [reply pageserver/src/basebackup.rs:324]: Yes, I am going to remove storing `pg_logical/replorigin_checkpoint` in future.
But as you have noticed, we need it for backward compatibility.
Also I do not want to include in this PR changes in Postgres code.

I will update comment.

Comment (knizhnik) [reply pageserver/src/basebackup.rs]: I have added more comments and add tests (it reproduced the problem on unpatched Neon).

File: pageserver/src/pgdatadir_mapping.rs
Comment (hlinnaka) [inline pageserver/src/pgdatadir_mapping.rs:1107]: If the `self.tline.repl_origins` cache isn't loaded yet, the `get_or_insert_with(ReplOrigins::default)` call will replace it with an empty value, losing all the other replication origins.

(With the approach of separate key-value pair for each replication origin, we don't need the cache at all, so this will become moot)

File: libs/pageserver_api/src/key.rs
Comment (hlinnaka) [reply libs/pageserver_api/src/key.rs:232]: cc @skyzh. Note that this is different from storing the replication slots, which previously caused the massive write amplification issues. These replication origins are much simpler, but I can see similar danger here too.

**Comment by skyzh** [comment]
This file is included in https://github.com/neondatabase/neon/blob/5a3d8e75edd5f684726b662638c833f02b1423e6/pageserver/src/aux_file.rs#L41

**Comment by skyzh** [comment]
okay, it seems that this file is not directly generated from Postgres but constructed on the page server. I'll need to revisit it and add it to aux v2.

File: libs/pageserver_api/src/key.rs
Comment (knizhnik) [reply libs/pageserver_api/src/key.rs:232]: > The max number of slots you can have, if you bump up max_replication_slots, is 2^16.
Actually larger - `MAX_BACKENDS = 0x3FFFF`.
But please notice that still real number of origins is limited by number of replicas.
Even if you bump up `max_replication_slots`, still you will not get more different origins than LR channels.
And number of this channels is limited by max_connection, which is 100, but certainly can be increased for some customers.

In any case, I am 100% for doing it properly, especially after this amplification problem with replslots.
But I am still not quite sure what properly means in this case.
What is the status of AUX2 files support now?
As far as I see, it is not yet merged. If we do not have method for range search of our KV storage, then I do not see some better alternative to the current approach (with  ReplOriginDir).

If we still have some kind of "directory" object containing IDs of all entries (as we have for databases, relations, ...), 
then IMHO there is no sense in storing separate entries for replorigins, because size of content (origin_lsn = 8bytes) is comparable with ID (origin_id - 4 bytes).

And if we really want to use range scan for replorigins, then it seems to me that:
1. It is better to wait until AUX2 stuff will be committed (if it really implements range scan).
2. Think about refactoring of relation directory, because it seems to suffer from the same amplification problem much more than replorigins.

And last one my comment/concern about "proper implementation": if we will use range scan, then it means that we have to reconstruct N objects instead of 1. It obviously increase time of preparing of basebackup and so startup time. So there is obvious compromise between space (write amplification) and extraction speed. I think that we should take it in account choosing right approach.

Comment (skyzh) [reply libs/pageserver_api/src/key.rs:232]: There is range scan. No directory keys. The last pull request in the queue will be merged today and I’ll have some perf improvement pull requests.

Comment (knizhnik) [reply libs/pageserver_api/src/key.rs]: I have changed this PR to store repl origins individually. and use range scan to fetch them.
@skyzh I will be pleased it you can review my changes (at least related with storing/retriving replorigins in KV storage). May be I missed something in how range scan works...

File: test_runner/regress/test_subscriber_restart.py
Comment (kelvich) [inline test_runner/regress/test_subscriber_restart.py:38]: will that test fail with the current implementation?

**Comment by skyzh** [comment]
Quick comment: the new keyspace seems not added to `collect_keyspace`, which indicates that compaction will drop the data. Apart from that, I'll do a thorough review likely next week.

File: test_runner/regress/test_subscriber_restart.py
Comment (knizhnik) [reply test_runner/regress/test_subscriber_restart.py:38]: Yes, it did. But not always.
I try to rewrite test to make it failed more reliably.

**Comment by knizhnik** [comment]
> Reference

Thank you, fixed.

File: libs/pageserver_api/src/key.rs
Comment (jcsp) [inline libs/pageserver_api/src/key.rs:603]: Can you use RepOriginId::MAX here?  (if the FFI declares it as a u16)

Or some other way make it clear here that 0x10000 is used because RepOriginId is a 16 bit type: otherwise it looks like a magic number.

File: test_runner/regress/test_subscriber_restart.py
Comment (jcsp) [inline test_runner/regress/test_subscriber_restart.py:7]: Can you write a docstring on this function that explains the purpose of the test -- as written, it's not really obvious that this is about logical replication unless the reader happens to know that "subscriber" means logical replication subscriber.

**Comment by andreasscherbaum** [comment]
This PR currently only tests restart of replication, which is an expected case.

It also needs tests for unexpected cases, like:
- invalid WAL
- wrong checksums
- unknown WAL entries
- ...

**Comment by knizhnik** [comment]
> This PR currently only tests restart of replication, which is an expected case.
> 
> It also needs tests for unexpected cases, like:
> 
> * invalid WAL
> * wrong checksums
> * unknown WAL entries
> * ...

This PR stores replication origins which are needed to correctly restart replication.
This is why test in this PR performs restarts of subscriber under workload.

All other reasons which you have mention have completely no relation to this PR.

File: libs/pageserver_api/src/key.rs
Comment (knizhnik) [reply libs/pageserver_api/src/key.rs:603]: It is exclusive boundary. So we can write it as `RepOrigin::MAX::uint32+1` but IMHO 0x10000 looks much more readable:)
In any case, my intention was to make range as small as possible, because I saw code pin PS which iterates through all keys in their range. But @VladLazar  said that it is not used in this case. So may be it is better to keep original range with `field1: REPL_ORIGIN_KEY_PREFIX+1`

Comment (VladLazar) [reply libs/pageserver_api/src/key.rs:603]: > because I saw code pin PS which iterates through all keys in their range

Just to confirm. `Timeline::scan` doesn't do that

**Comment by skyzh** [comment]
I still have several pull requests targeting aux files waiting for merge before I could test + benchmark this one. Sorry for that and I will probably do the reviews next week :) I will also help resolve conflicts and help with the tests :)

**Comment by skyzh** [comment]
Resolved merge conflicts. Note that there is an assertion on the freeze in memory layer path that asserts metadata keyspace contains only one partition. I created https://github.com/neondatabase/neon/pull/7865 to support multiple partitions, because this pull request requires two keyspaces in the metadata keys.

**Comment by kelvich** [comment]
@knizhnik @skyzh now pageserver tests are erroring out:

```
--- TRY 4 STDERR:        pageserver tenant::storage_layer::layer::tests::layer_size ---
thread 'tenant::storage_layer::layer::tests::layer_size' panicked at pageserver/src/tenant/storage_layer/layer/tests.rs:819:5:
assertion `left == right` failed
```

**Comment by skyzh** [comment]
This test is added to ensure people don't add too large fields to the layer descs. Need to change the value in that test to match the latest fields.

File: libs/pageserver_api/src/key.rs
Comment (skyzh) [reply libs/pageserver_api/src/key.rs:603]: I'm fine with a smaller range right now. The storage engine does not iterate all keys for sparse keyspaces.

**Review by skyzh — APPROVED** [review]
LGTM for the storage part of this pull request. Added a few commits to ensure storage things work. The new integration test seems flaky. I think we need to make it work + a compute team's approval before merging.

**Review by save-buffer — APPROVED** [review]

**Pull request merged**
